include/my_sys.h

```diff
@@ -921,16 +921,31 @@ extern int my_compress_buffer(uchar *dest, size_t *destLen,
 extern int packfrm(const uchar *, size_t, uchar **, size_t *);
 extern int unpackfrm(uchar **, size_t *, const uchar *);
 
-extern ha_checksum my_checksum(ha_checksum crc, const uchar *mem,
-                               size_t count);
+void my_checksum_init(void);
+#ifdef HAVE_CRC32_VPMSUM
+extern my_checksum(ha_checksum, const void *, size_t);
+#else
+typedef ha_checksum (*my_crc32_t)(ha_checksum, const void *, size_t);
+extern my_crc32_t my_checksum;
+#endif
+
+#if __GNUC__ >= 4 && defined(__x86_64__)
+int crc32_pclmul_enabled();
+unsigned int crc32_pclmul(unsigned int, const void *, size_t);
+#elif defined(__GNUC__) && defined(__linux__) && defined(HAVE_ARMV8_CRC)
+#include <stdint.h>
+unsigned int crc32_aarch64_available(void);
+unsigned int crc32_aarch64(unsigned int, const void *, size_t);
+uint32_t crc32c_aarch64(uint32_t, const unsigned char *, uint64_t);
+#endif
```

Comment:
Do we really have to declare `crc32_pclmul_enabled()`, `crc32_pclmul()`, `crc32_aarch64()` here? They should never be called directly, only via `my_checksum_init()` or the `my_checksum` function pointer. Ideally, these would be declared `static` in `mysys/checksum.c` which seems to be the only compilation unit where `my_checksum_init()` is defined. If the functions must be defined in separate compilation units, then I think that it would be better that `mysys/checksum.c` includes implementation-specific header files that declares those functions.

Is the `defined(__linux__)` really necessary? I could imagine that the same code and toolchain would work on (say) FreeBSD or Debian/kFreeBSD.

I do not think that the declarations of `crc32_pclmul()` or `crc32_aarch64()` belong here. The functions should never be called directly, always only via the `my_checksum` function pointer. But I can accept them if we cannot 

I understand that you are declaring `crc32_aarch64_available()` and `crc32c_aarch64()` because they are used by `ut0crc32.cc`. That is fine until a future effort cleans up the InnoDB `ut_crc32()`. But, doing `#include <stdint.h>` for only one platform feels very wrong.

I think that all occurrences of `unsigned int` or `uint32_t` in the above snippet should be replaced with `ha_checksum`. And I think that only the following should remain:
```c
#ifdef HAVE_ARMV8_CRC
int crc32_aarch64_available(void);
ha_checksum crc32c_aarch64(ha_checksum, const void *, size_t);
#endif
```
We customarily use the `int` return type for Boolean predicates in C code, instead of `unsigned int`.

Yes, I would suggest the types of the second and third parameter to change. (It should be no real change, because `sizeof(size_t)==sizeof(uint64_t)` on ARMv8.) And I would expect that the compile-time check of `HAVE_ARMV8_CRC` to fail if a `__GNUC__` compatible compiler (GCC or clang) is not available.

Reply:
If you do not want to introduce implementation-specific header files that would only be `#include`d from `mysys/checksum.c`, then you could declare those implementation-specific functions in `mysys/checksum.c` only, similar to how `storage/innobase/ut/ut0crc32.cc` was doing it.

Reply:
ok .. so function are being used only in checksum.c so moved their declaration back to checksum.c. crc32_aarch64_available is used from checksum.c and ut0crc32 so as per original comment left it out in my_sys to avoid C declaration mismatch.